src/components/tree-view/tree-view.tsx

```diff
@@ -25,37 +28,77 @@ export interface TreeViewProps {
     selectedItem?: TreeItemData;
 }
 
+export interface TreeItemState {
+    isSelected: boolean;
+}
+
+export type StateMap = Map<TreeItemData, TreeItemState>;
+
 const itemIdPrefix = 'TREE_ITEM';
 
-export function TreeItem({ item, itemRenderer, onItemClick, isSelected }: TreeItemProps): JSX.Element {
+export function TreeItem({ item, itemRenderer, onItemClick, stateMap, state }: TreeItemProps): JSX.Element {
+    const itemLabel = item.label.replace(' ', '_');
     return (
-        <div key={item.label}>
-            <div data-automation-id={`${itemIdPrefix}_${item.label.replace(' ', '_')}`} className={style['tree-node']}
-                 onClick={() => onItemClick!(item)} data-selected={isSelected(item)}>
-                <span data-automation-id={`${itemIdPrefix}_${item.label}_ICON`}>&gt; </span>
-                <span data-automation-id={`${itemIdPrefix}_${item.label}_LABEL`}>{item.label}</span>
+        <div>
+            <div data-automation-id={`${itemIdPrefix}_${itemLabel}`} className={style['tree-node']}
+                 onClick={() => onItemClick!(item)} data-selected={ state!.isSelected }>
+                <span data-automation-id={`${itemIdPrefix}_${itemLabel}_ICON`}>&gt; </span>
+                <span data-automation-id={`${itemIdPrefix}_${itemLabel}_LABEL`}>{item.label}</span>
             </div>
             <div className={style['nested-tree']}>
-                {(item.children || []).map((child: TreeItemData) =>
-                    itemRenderer({item: child, onItemClick, itemRenderer, isSelected}))}
+                {(item.children || []).map((child: TreeItemData, index: number) =>
+                    React.createElement(itemRenderer,
+                        {item: child, onItemClick, itemRenderer, stateMap, state: stateMap.get(child)!, key: `${index}`}))}
             </div>
         </div>
     )
 }
 
+const TreeItemWrapper = observer(TreeItem);
+
+@observer
 export class TreeView extends React.Component<TreeViewProps, {}>{
-    static defaultProps = { itemRenderer: TreeItem, onSelectItem: () => {} };
+    static defaultProps = { itemRenderer: TreeItemWrapper, onSelectItem: () => {} };
+
+    stateMap: StateMap = new Map<TreeItemData, TreeItemState>();
+
+    constructor(props: TreeViewProps) {
+        super(props);
+        this.initStateMap(props.dataSource);
+    }
+
+    initStateMap(data: Object[] = []) {
+        data.forEach((item: TreeItemData) => {
+            this.stateMap.set(item, observable({ isSelected: false }));
```

Comment:
need `observable` elements, not the map itself.
also, this map states `only strings are accepted as keys.` while we need to use object references as keys.